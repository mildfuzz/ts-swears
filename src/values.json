{
  "ddl": {
    "page": {
      "nbs_environment": null,
      "nbs_server": null,
      "pageInfo": {
        "pageName": null,
        "nbs_page_load_time": null,
        "nbs_page_responsive_state": null,
        "nbs_page_responsive_orientation": null
      },
      "category": {
        "primaryCategory": "over payment application",
        "nbs_sub_category_1": null,
        "nbs_sub_category_2": null,
        "nbs_sub_category_3": null
      }
    },

    "nbs_message": [],
    "nbs_user_input": [],
    "nbs_element_interaction": [],
    "nbs_app_mortgage": {},

    "nbs_user": {
      "nbs_user_application_route": null
    }
  },

  "regionProperties": {
    "view": ["page"],
    "action": ["nbs_user_input", "nbs_element_interaction", "nbs_message"],
    "context": ["nbs_user", "nbs_app_mortgage"]
  },

  "itemsKeyMap": {
    "nbs_message": {
      "category": null,
      "text": null,
      "cause": null
    },

    "nbs_user_input": {
      "name": "nbs_input_name",
      "status": "nbs_input_validation_status",
      "message": "nbs_input_validation_message"
    },

    "nbs_element_interaction": {
      "type": "nbs_interaction_type",
      "label": "nbs_interaction_label"
    }
  },

  "page": {
    "PREFIX": "nbs-aws",
    "TYPE": "application",
    "APP_NAME": "OverPayment",
    "INITIAL": "initial load",

    "types": {
      "LOADER": "loader"
    },
    "orientations": {
      "PORTRAIT": "portrait",
      "LANDSCAPE": "landscape",
      "SQUARE": "square"
    },
    "sizes": {
      "SMALL": "small",
      "MEDIUM": "medium",
      "LARGE": "large"
    }
  },

  "valueMaps": {
    "environments": {
      "home-mumo-dev": "dev",
      "home-mumo-prd": "live",
      "home-afford-dev-rtl": "dev",
      "home-afford-sit-rtl": "sit"
    },

    "routes": {
      "referrers": {
        "://(www.)?nationwide.co.uk": "nationwide.co.uk",
        "://onlinebanking.nationwide.co.uk": "internet bank",
        "^$": "external"
      },
      "urlParams": {
        "JId": "banking app"
      }
    },

    "inputValidity": {
      "true": "passed",
      "false": "failed",
      "null": "none"
    }
  },

  "types": {
    "action": {
      "SUBMIT": "submit",
      "INPUT": "user_input_complete",
      "DEFAULT": "misc_interaction"
    },

    "interaction": {
      "?LINK": "external link",
      "SUBMIT": "submit",
      "REVEAL": "reveal",
      "HELP": "help icon",
      "PRINT": "print",
      "PDF": "save pdf",
      "CALL": "click to call",
      "BACK": "back",
      "MISC": "misc"
    },

    "message": {
      "USER": "user error",
      "BUSINESS": "business error",
      "SYSTEM": "system error",
      "NOTIFICATION": "notification"
    }
  }
}
