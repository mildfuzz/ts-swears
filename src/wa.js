/* eslint-disable no-console */
import _cloneDeep from 'lodash/cloneDeep'
import _isFunction from 'lodash/isFunction'
/**
 * Web Analytic Functions
 */
export const safely = (fn, name, ...args) => {
  if (_isFunction(fn)) {
    try {
      fn(...args)
      console.log(name, ...args)
    } catch (err) {
      console.error(err)
    }
  } else {
    console.error(name, 'not available')
  }
}
const wa = {
  view: null,
  action: null,
  restore: null,
  backup: null,
}
const isDev = process.env.NODE_ENV === 'development'

// const isDev = false
if (isDev) {
  let ddlBackup = null
  wa.view = () => console.info('wa_view', _cloneDeep(global.digitalData))
  wa.action = (...args) => {
    console.info('wa_action', _cloneDeep(global.digitalData), ...args)
  }
  wa.ddlRestore = () => {
    global.digitalData = ddlBackup
    ddlBackup = null
    console.info('ddl_restore', _cloneDeep(global.digitalData))
  }
  wa.ddlBackup = () => {
    ddlBackup = _cloneDeep(global.digitalData)
    console.info('ddl_backup', ddlBackup)
  }
  global.digitalData = {}
  console.log('DDL created', !!global.digitalData)
} else {
  wa.view = (...args) => safely(global.wa_view, 'wa_view', ...args)
  wa.action = (...args) => safely(global.wa_action, 'wa_action', ...args)
  wa.ddlBackup = (...args) => safely(global.ddl_backup, 'ddl_backup', ...args)
  wa.ddlRestore = (...args) =>
    safely(global.ddl_restore, 'ddl_restore', ...args)

  global.digitalData = {}
}
export default wa
