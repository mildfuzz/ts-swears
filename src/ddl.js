import _get from 'lodash/get'
import _set from 'lodash/set'
import _pick from 'lodash/pick'
import _mapKeys from 'lodash/mapKeys'
import _cloneDeep from 'lodash/cloneDeep'
import _defaultsDeep from 'lodash/defaultsDeep'
import _uniqBy from 'lodash/uniqBy'

import * as utils from './utils'
import values from './values.json'

const {valueMaps, regionProperties} = values
const getInitialViewData = () =>
  _defaultsDeep(
    {},
    _pick(values.ddl, regionProperties.view, regionProperties.action),
  )

const contextData = _defaultsDeep(
  {},
  _pick(values.ddl, regionProperties.context),
)
let viewData = getInitialViewData()

/**
 * Internal API
 */
const ddl = {
  /**
   * Get property data source
   * @param {string} property
   */
  getSource: (property) => {
    if (utils.isContextProperty(property)) return contextData
    if (utils.isActionProperty(property) || property === 'page')
      return global.digitalData
    return viewData
  },

  /**
   * Get data from ddl
   * @param {string} property
   */
  get: (property) => _get(ddl.getSource(property), property),

  /**
   * Create new ddl
   * @param {object} data
   */
  create: (data) => {
    global.digitalData = _cloneDeep(
      _defaultsDeep(data, _cloneDeep(contextData), viewData),
    )
    viewData = getInitialViewData()
  },

  /**
   * Merge contextual data with ddl
   */
  refresh: () => {
    global.digitalData = _defaultsDeep(
      {},
      _cloneDeep(contextData),
      global.digitalData,
    )
  },

  /**
   * Update existing ddl data
   * @param {string} property
   * @param {*} data
   */
  update: (property, newData, applyDefaults = true) => {
    const source = ddl.getSource(property)
    const data = applyDefaults
      ? _defaultsDeep(_cloneDeep(newData), ddl.get(property))
      : _cloneDeep(newData)
    _set(source, property, data)
  },

  /**
   * Add ddl item to a ddl array - new items are always at head
   * @param {string} itemProperty
   * @param {*} item
   */
  add: (itemProperty, item, applyDefaults = true) => {
    const ddlItems = ddl.get(itemProperty) || []
    const ddlItem = _mapKeys(item, (_, property) =>
      utils.getItemPropertyKey(itemProperty, property),
    )

    ddl.update(itemProperty, [ddlItem, ...ddlItems], applyDefaults)
  },

  /**
   * Add ddl item to a ddl array - new items are always at head
   * @param {string} itemProperty
   * @param {*} item
   */
  replace: (itemProperty, item, applyDefaults = true) => {
    const ddlItems = ddl.get(itemProperty) || []
    const ddlItem = _mapKeys(item, (_, property) =>
      utils.getItemPropertyKey(itemProperty, property),
    )

    const messages = _uniqBy([ddlItem, ...ddlItems], 'cause')

    ddl.update(itemProperty, messages, applyDefaults)
  },
}

export default {
  /**
   * spec - 5.1.6.1
   * @param {string|array} desc
   */
  ...ddl,

  setPage: (desc, isProd = false) => {
    const {getPageName} = utils
    const pageName = getPageName(desc)
    const category = ''

    ddl.create({
      page: {
        pageInfo: {
          pageName,
          nbs_page_load_time: utils.loadEnd(pageName),
          ...utils.getDeviceData({
            width: global.innerWidth,
            height: global.innerHeight,
          }),
        },
        category,
        ...utils.getServiceData(isProd ? 'live' : 'dev'),
      },
    })
  },

  updatePageName: (desc) => {
    const {getPageName} = utils
    const pageName = getPageName(desc)
    const category = ''

    ddl.update('page', {
      pageInfo: {
        pageName,
      },
      category,
    })

    ddl.refresh()
  },

  setUserApplicationRoute: ({applicationRoute}) => {
    ddl.update('nbs_user', {
      nbs_user_application_route: utils.getDdlText(applicationRoute),
    })
  },

  /**
   * Spec: 5.1.6.5
   */
  addMessage: (category, text, cause) => {
    ddl.add('nbs_message', {
      category: utils.getDdlText(category),
      text: utils.getDdlText(text),
      cause: utils.getDdlText(cause),
    })
  },

  replaceMessage: (category, text, cause) => {
    ddl.replace('nbs_message', {
      category: utils.getDdlText(category),
      text: utils.getDdlText(text),
      cause: utils.getDdlText(cause),
    })
  },

  /**
   * Spec: 5.2.1.2
   */
  addUserInput: (name, message, isValid = null, applyDefaults = true) => {
    ddl.add(
      'nbs_user_input',
      {
        name: utils.getDdlText(name),
        message: utils.getDdlText(message),
        status: valueMaps.inputValidity[isValid],
      },
      applyDefaults,
    )
  },

  /**
   * Spec: 5.2.X.3
   */
  addInteraction: (type, label) => {
    ddl.add('nbs_element_interaction', {
      type: utils.getDdlText(type),
      label: utils.getDdlText(label),
    })
  },

  updateMortgageObject: (analyticsValues) =>
    ddl.update('nbs_app_mortgage', analyticsValues),
  /**
   * Additional utilities
   */
  refresh: () => ddl.refresh(),
}
