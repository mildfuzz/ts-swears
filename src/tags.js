import _size from 'lodash/size'
import _reduce from 'lodash/reduce'
import _flatten from 'lodash/flatten'
import _isEmpty from 'lodash/isEmpty'
import _memoize from 'lodash/memoize'
import _mapKeys from 'lodash/mapKeys'
import _snakeCase from 'lodash/snakeCase'
import _replace from 'lodash/replace'
import _defaults from 'lodash/defaults'
import _camelCase from 'lodash/camelCase'
import _isUndefined from 'lodash/isUndefined'

import values from './values.json'
import ddl from './ddl'

import wa from './wa'

/**
 * Web Analytic Scenarios
 */
export const tags = {
  /**
   * Spec: 5.1.2 & 5.1.3 - Standard / Dynamic Page View
   */
  pageView: ({name}) => {
    ddl.setPage(name)
    wa.view()
  },

  pageNameUpdateView: (desc) => {
    ddl.updatePageName(desc)
    wa.view()
  },

  /**
   * Spec: 5.1.4 - Overlay Page View (Trigger)
   */
  overlayOpen: ({name, ...opts}) => {
    wa.ddlBackup()

    tags.pageView({
      ...opts,
      name: _flatten([name]),
    })
  },

  modalOpen: (opts) => tags.overlayOpen(opts),
  loaderOpen: (opts) => tags.overlayOpen(opts),
  sliderOpen: (opts) => tags.overlayOpen(opts),

  /**
   * Spec: 5.1.5 - Overlay Page View (Close)
   */
  overlayClose: ({isReturning = true} = {}) => {
    wa.ddlRestore()

    setTimeout(() => {
      // set timeout enables support for ddl updates before refresh & page view
      ddl.refresh()

      if (isReturning === true) {
        wa.view()
      }
    })
  },

  modalClose: (opts) => tags.overlayClose(opts),
  loaderClose: (opts) => tags.overlayClose({...opts, isReturning: false}),
  sliderClose: (opts) => tags.overlayClose(opts),
  /**
   * Spec: 5.1.6.5 - Error & Notification Messages
   */
  ..._reduce(
    values.types.message,
    (messageTags, type) => ({
      ...messageTags,
      [_camelCase(`${type} message`)]: (message, cause) =>
        ddl.addMessage(type, message, cause),
    }),
    {},
  ),

  /**
   * Spec: 5.2.1.1 - User Input
   */
  userInput: ({label, message, isValid}) => {
    ddl.addUserInput(label, message, isValid, false)
    wa.action(values.types.action.INPUT)
  },

  /**
   * Spec: 5.2 - Page Interactions
   */
  ..._reduce(
    values.types.interaction,
    (interactionTags, type, interaction) => ({
      ...interactionTags,
      [_camelCase(`${interaction} interaction`)]: (label) => {
        ddl.addInteraction(type, label)
        wa.action(
          values.types.action[interaction] || values.types.action.DEFAULT,
        )
      },
    }),
    {},
  ),

  /**
   * This submit interaction handles NEL standard behaviour
   * If there are form validation messages then these will be added as user error messages
   * In addition, if the submit is invalid, as controlled by the valid argument, which defaults to
   * the presence of error messages, then wa_view will again be called so that analytics is aware
   * that it is on the same view and has not progressed to the next view
   *
   * the cause would likely be the field name
   */
  submitInteractionWithValidation: ({
    label,
    messages = [],
    valid,
    analyticsValues,
  }) => {
    // add basic submit interaction
    ddl.addInteraction(values.types.action.SUBMIT, label)

    if (!_isEmpty(analyticsValues)) {
      tags.sendValuesToAdobe(analyticsValues)
    }

    const hasMessages = !!_size(messages)

    //  add any validation messages

    messages.forEach(({message, cause}) => {
      ddl.replaceMessage(values.types.message.USER, message, cause)
    })

    wa.action('form_submission')

    // determine whether should fire wa_view, normally because submit operation failed
    const shouldRefreshView = _isUndefined(valid) ? hasMessages : !valid

    if (shouldRefreshView) {
      wa.view()
    }
  },

  /**
   * If field values are to be added for the analytics then use this function
   * @param {object} vals the values to be added to the ddl object
   * @param {string[]} accessors the fields to be added to the ddl object. They use `_get` for accessing nested properties
   *
   * @example
   * sendValuesToAdobe({
   *  mortgageLength: { years: 10, months: 5 },
   *  ltv: 70,
   *  loanAmount: 150000
   *  })
   */
  sendValuesToAdobe: (vals) => {
    const tagVals = _mapKeys(
      vals,
      (_, key) => _snakeCase(_replace(key, /\./g, ' ')),
      {},
    )

    ddl.updateMortgageObject(tagVals)
  },
}

/**
 * Spec: 5.2.4.1 - Condition 1 (only tag first interaction)
 */
tags.helpInteraction = _memoize(tags.helpInteraction)

/**
 * Spec: 5.2.3.1 - Fig. 7 (only tag reveal)
 */
const {revealInteraction} = tags
tags.revealInteraction = (label, isHiding) =>
  !isHiding && revealInteraction(label)

/**
 * Create view tag handler
 * @param {function} type
 * @param {object} opts
 */
const createViewTagHandler = (scenarioTag, tagOptions) => {
  const tagHandler = (opts) => {
    scenarioTag(_defaults(opts, tagOptions))
  }

  tagHandler.tagOptions = tagOptions

  return tagHandler
}

/**
 * Create view tags handler functions from scenario tags
 * @param {object} tagScenarios
 * @returns {object} tags
 */
export const createViewTags = (tagScenarios) =>
  _reduce(
    tagScenarios,
    (viewTags, {type, ...opts}, name) => ({
      ...viewTags,
      [name]: createViewTagHandler(type, opts),
    }),
    {},
  )

/**
 * Wrapper got link interaction, workaround for TS checking
 * as TS can't infer type from dynamically interactions
 * @param {string} message
 */

export const handleLinkInteraction = (message) => {
  tags.linkInteraction(message)
}
