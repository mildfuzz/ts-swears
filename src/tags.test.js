import _pick from 'lodash/pick'
import _forEach from 'lodash/forEach'
import _camelCase from 'lodash/camelCase'

import values from './values.json'
import ddl from './ddl'
import tags, {createViewTags} from './tags'
import wa from './wa'

jest.mock('./ddl')

describe('@analytics > tags', () => {
  jest.useFakeTimers()

  jest.spyOn(wa, 'view').mockImplementation()
  jest.spyOn(wa, 'ddlBackup').mockImplementation()
  jest.spyOn(wa, 'ddlRestore').mockImplementation()
  jest.spyOn(wa, 'action').mockImplementation()

  beforeEach(() => jest.clearAllMocks())

  describe('tagging scenarios', () => {
    test('api', () => {
      expect(tags).toMatchSnapshot()
    })

    describe('view tagging', () => {
      _forEach(
        [
          {tagType: 'pageView', providedName: ['foo', 'bar']},
          {
            tagType: 'loaderOpen',
            providedName: ['foo', 'bar'],
            expectedName: ['foo', 'bar'],
          },
          {
            tagType: 'overlayOpen',
            providedName: ['foo', 'bar'],
            expectedName: ['foo', 'bar'],
          },
          {
            tagType: 'modalOpen',
            providedName: ['foo', 'bar'],
            expectedName: ['foo', 'bar'],
          },
        ],
        ({tagType, providedName, expectedName = providedName}) => {
          describe(`${tagType}()`, () => {
            test('call ddl set page with initial page description', () => {
              tags[tagType]({name: providedName})
              jest.runOnlyPendingTimers()
              expect(ddl.setPage).toHaveBeenCalledTimes(1)
              expect(ddl.setPage).toHaveBeenCalledWith(expectedName)
            })

            test('calls web analytics view', () => {
              tags[tagType]({name: providedName})
              jest.runOnlyPendingTimers()
              expect(wa.view).toHaveBeenCalledTimes(1)
            })
          })
        },
      )

      describe('pageNameUpdateView()', () => {
        const updatedPageName = ['bar 1', 'bar 2']

        test('call ddl update page name with page description', () => {
          tags.pageNameUpdateView(updatedPageName)

          expect(ddl.updatePageName).toHaveBeenCalledTimes(1)
          expect(ddl.updatePageName).toHaveBeenCalledWith(updatedPageName)
        })

        test('calls web analytics view', () => {
          tags.pageNameUpdateView(updatedPageName)

          expect(wa.view).toHaveBeenCalledTimes(1)
        })
      })
    })

    describe('overlay / modal tagging', () => {
      _forEach(['overlay', 'modal', 'loader', 'slider'], (tagType) => {
        describe(`${tagType}Open()`, () => {
          test('calls web analytics ddl backup and view', () => {
            tags[`${tagType}Open`]({})
            jest.runOnlyPendingTimers()
            expect(wa.ddlBackup).toHaveBeenCalledTimes(1)
            expect(wa.view).toHaveBeenCalledTimes(1)
          })
        })

        describe(`${tagType}Close()`, () => {
          test('calls web analytics ddl restore', () => {
            const closeTag = tags[`${tagType}Close`]
            closeTag()
            jest.runOnlyPendingTimers()
            expect(wa.ddlRestore).toHaveBeenCalledTimes(1)
          })

          if (tagType !== 'loader') {
            test('calls refresh and only call web analytics view if returning to page', () => {
              const closeTag = tags[`${tagType}Close`]

              closeTag()
              jest.runOnlyPendingTimers()
              closeTag({isReturning: true})
              jest.runOnlyPendingTimers()
              closeTag({isReturning: false})
              jest.runOnlyPendingTimers()

              expect(ddl.refresh).toHaveBeenCalledTimes(3)
              expect(wa.view).toHaveBeenCalledTimes(2)
            })
          }
        })
      })
    })

    describe('message tagging', () => {
      _forEach(values.types.message, (messageType) => {
        describe(`${_camelCase(messageType)}Message()`, () => {
          test('calls ddl add message', () => {
            const message = `${messageType} type message`
            const cause = `${messageType} type cause`

            tags[`${_camelCase(messageType)}Message`](message, cause)
            jest.runOnlyPendingTimers()
            expect(ddl.addMessage).toHaveBeenCalledTimes(1)
            expect(ddl.addMessage).toHaveBeenCalledWith(
              messageType,
              message,
              cause,
            )
          })
        })
      })
    })

    describe('action tagging', () => {
      describe('userInput()', () => {
        test('calls ddl add user input', () => {
          const label = 'input label'
          const message = 'input validation message'
          const isValid = null
          const applyDefaults = false

          tags.userInput({label, message, isValid})
          jest.runOnlyPendingTimers()
          expect(ddl.addUserInput).toHaveBeenCalledTimes(1)
          expect(ddl.addUserInput).toHaveBeenCalledWith(
            label,
            message,
            isValid,
            applyDefaults,
          )
        })

        test('calls web analytics action with event', () => {
          tags.userInput({})
          jest.runOnlyPendingTimers()
          expect(wa.action).toHaveBeenCalledTimes(1)
          expect(wa.action).toHaveBeenCalledWith(values.types.action.INPUT)
        })
      })
    })

    describe('interaction tagging', () => {
      _forEach(values.types.interaction, (interactionType, interactionName) => {
        describe(`${_camelCase(interactionName)}Interaction()`, () => {
          const tagInteraction =
            tags[`${_camelCase(interactionName)}Interaction`]

          test('calls ddl add interaction', () => {
            const label = `${interactionType} type label`

            tagInteraction(label)
            jest.runOnlyPendingTimers()
            expect(ddl.addInteraction).toHaveBeenCalledTimes(1)
            expect(ddl.addInteraction).toHaveBeenCalledWith(
              interactionType,
              label,
            )
          })

          test('calls web analytics action with event', () => {
            tagInteraction()
            jest.runOnlyPendingTimers()
            expect(wa.action).toHaveBeenCalledTimes(1)
            expect(wa.action).toHaveBeenCalledWith(
              values.types.action[interactionName] ||
                values.types.action.DEFAULT,
            )
          })

          if (interactionName === 'HELP') {
            test('only tag first help interaction', () => {
              tagInteraction('a')
              tagInteraction('b')
              tagInteraction('a')
              tagInteraction('a')
              tagInteraction('b')
              tagInteraction('b')
              jest.runOnlyPendingTimers()
              expect(ddl.addInteraction).toHaveBeenCalledTimes(2)
              expect(wa.action).toHaveBeenCalledTimes(2)
            })
          }

          if (interactionName === 'REVEAL') {
            test('only tag reveal interaction on reveal', () => {
              tagInteraction('a', false)
              tagInteraction('b', false)
              tagInteraction('a', true)
              tagInteraction('b', true)
              jest.runOnlyPendingTimers()
              expect(ddl.addInteraction).toHaveBeenCalledTimes(2)
              expect(wa.action).toHaveBeenCalledTimes(2)
            })
          }
        })
      })
    })

    describe('submitInteractionWithValidation()', () => {
      const label = 'submit button'
      test('calls ddl add interaction with submit tyoe and label', () => {
        tags.submitInteractionWithValidation({label})
        jest.runOnlyPendingTimers()
        expect(ddl.addInteraction).toHaveBeenCalledTimes(1)
        expect(ddl.addInteraction).toHaveBeenCalledWith(
          values.types.action.SUBMIT,
          label,
        )
      })

      test('calls wa_action after adding ddl interaction', () => {
        tags.submitInteractionWithValidation({label})
        jest.runOnlyPendingTimers()
        expect(wa.action).toHaveBeenCalledTimes(1)
        expect(wa.action).toHaveBeenCalledWith('form_submission')
        expect(wa.view).not.toHaveBeenCalled()
      })

      test('adds messages if present and defaults to invalid with messages', () => {
        const messages = [
          {cause: 'firstName', message: 'This is a required field'},
          {cause: 'email', message: 'This is not a valid email'},
        ]
        tags.submitInteractionWithValidation({label, messages})
        jest.runOnlyPendingTimers()
        expect(ddl.replaceMessage).toHaveBeenCalledTimes(2)
        expect(ddl.replaceMessage).toHaveBeenCalledWith(
          values.types.message.USER,
          'This is a required field',
          'firstName',
        )
        expect(ddl.replaceMessage).toHaveBeenCalledWith(
          values.types.message.USER,
          'This is not a valid email',
          'email',
        )
        expect(wa.view).toHaveBeenCalledTimes(1)
      })

      test('adds messages if present and accepts also a valid argument which prevents wa_view being called if true', () => {
        const messages = [
          {cause: 'firstName', message: 'This is a required field'},
          {cause: 'email', message: 'This is not a valid email'},
        ]
        tags.submitInteractionWithValidation({label, messages, valid: true})
        jest.runOnlyPendingTimers()
        expect(ddl.replaceMessage).toHaveBeenCalledTimes(2)
        expect(ddl.replaceMessage).toHaveBeenCalledWith(
          values.types.message.USER,
          'This is a required field',
          'firstName',
        )
        expect(ddl.replaceMessage).toHaveBeenCalledWith(
          values.types.message.USER,
          'This is not a valid email',
          'email',
        )
        expect(wa.view).toHaveBeenCalledTimes(0)
      })

      test('will also send values to adobe if an analytics object is passed in', () => {
        const analyticsValues = {
          mortgageLength: {
            years: 20,
            months: 6,
          },
          ltv: 66.66,
          purchasePrice: 300000,
          loanAmount: 200000,
        }
        tags.submitInteractionWithValidation({
          label,
          analyticsValues,
          valid: true,
        })
        jest.runOnlyPendingTimers()
        expect(wa.action).toHaveBeenCalledTimes(1)
        expect(wa.action).toHaveBeenCalledWith('form_submission')
        expect(wa.view).not.toHaveBeenCalled()
        expect(wa.view).toHaveBeenCalledTimes(0)

        expect(_pick(global.digitalData, 'nbs_app_mortgage')).toMatchSnapshot()
      })

      test('should call updateMortgageObject', () => {
        const analyticsValues = {
          mortgageLength: {
            years: 20,
            months: 6,
          },
          ltv: 66.66,
          purchasePrice: 300000,
          loanAmount: 200000,
        }
        tags.submitInteractionWithValidation({
          label,
          analyticsValues,
          valid: true,
        })

        expect(ddl.updateMortgageObject).toHaveBeenCalled()
      })
    })

    describe('sendValuesToAdobe()', () => {
      test('adds values to the ddl that matches the array of passed in strings', () => {
        const vals = {
          mortgageLength: {
            years: 20,
            months: 6,
          },
          ltv: 66.66,
          purchasePrice: 300000,
          loanAmount: 200000,
        }

        tags.sendValuesToAdobe(vals)

        expect(_pick(global.digitalData, 'nbs_app_mortgage')).toMatchSnapshot()
      })
    })
  })

  describe('createViewTagHandler()', () => {
    const tagScenarios = {
      foo: {
        type: jest.fn(),
        name: 'foo',
        optionA: [true, false],
        optionB: {foo: true},
      },
      bar: {type: jest.fn(), name: 'bar'},
      baz: {type: jest.fn(), name: 'baz'},
      quz: {type: jest.fn(), name: 'quz'},
    }
    const viewTags = createViewTags(tagScenarios)

    test('creates view tag handler functions from provided tag scenarios', () => {
      expect(viewTags).toMatchSnapshot()
    })

    describe('view tag', () => {
      test('exposes tag options', () => {
        _forEach(tagScenarios, ({type, ...opts}) => {
          expect(viewTags[opts.name].tagOptions).toMatchSnapshot()
        })
      })

      test('handler function calls tag type with tag options', () => {
        _forEach(tagScenarios, ({type, ...tagOptions}) => {
          const options = {ownOption: true}
          viewTags[tagOptions.name](options)
          expect(type).toHaveBeenCalledTimes(1)
          expect(type).toHaveBeenCalledWith({
            ...tagOptions,
            ...options,
          })
        })
      })
    })
  })
})
