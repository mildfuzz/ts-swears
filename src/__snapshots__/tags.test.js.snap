// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@analytics > tags createViewTagHandler() creates view tag handler functions from provided tag scenarios 1`] = `
Object {
  "bar": [Function],
  "baz": [Function],
  "foo": [Function],
  "quz": [Function],
}
`;

exports[`@analytics > tags createViewTagHandler() view tag exposes tag options 1`] = `
Object {
  "name": "foo",
  "optionA": Array [
    true,
    false,
  ],
  "optionB": Object {
    "foo": true,
  },
}
`;

exports[`@analytics > tags createViewTagHandler() view tag exposes tag options 2`] = `
Object {
  "name": "bar",
}
`;

exports[`@analytics > tags createViewTagHandler() view tag exposes tag options 3`] = `
Object {
  "name": "baz",
}
`;

exports[`@analytics > tags createViewTagHandler() view tag exposes tag options 4`] = `
Object {
  "name": "quz",
}
`;

exports[`@analytics > tags tagging scenarios api 1`] = `
Object {
  "backInteraction": [Function],
  "businessErrorMessage": [Function],
  "callInteraction": [Function],
  "helpInteraction": [Function],
  "linkInteraction": [Function],
  "loaderClose": [Function],
  "loaderOpen": [Function],
  "miscInteraction": [Function],
  "modalClose": [Function],
  "modalOpen": [Function],
  "notificationMessage": [Function],
  "overlayClose": [Function],
  "overlayOpen": [Function],
  "pageNameUpdateView": [Function],
  "pageView": [Function],
  "pdfInteraction": [Function],
  "printInteraction": [Function],
  "revealInteraction": [Function],
  "sendValuesToAdobe": [Function],
  "sliderClose": [Function],
  "sliderOpen": [Function],
  "submitInteraction": [Function],
  "submitInteractionWithValidation": [Function],
  "systemErrorMessage": [Function],
  "userErrorMessage": [Function],
  "userInput": [Function],
}
`;

exports[`@analytics > tags tagging scenarios sendValuesToAdobe() adds values to the ddl that matches the array of passed in strings 1`] = `Object {}`;

exports[`@analytics > tags tagging scenarios submitInteractionWithValidation() will also send values to adobe if an analytics object is passed in 1`] = `Object {}`;
